Async actor

stefanhp opened the pull request.

Commit:
Add first draft of AsyncActor with non working timeout

stefanhp committed:
build: async-actor feature, update to rust 1.75

- add an async-actor feature, enabled by default
- update minimum version of rust to 1.75 (async fn in trait)
- update default version of rust to 1.86
- update CI to use those rust versions
- update macro version to 0.5.2 to match main crate version
- fix a clippy warning in tests

stefanhp committed:
test: add tests for AsyncActor

- use a smol test executor for the test async runtime
- remove async from send_notification since it does not use async code

Note: reordered dependencies alphabetically in Cargo.toml

mrcmry commented on Cargo.toml:
```diff
@@ -6,20 +6,30 @@ license = "MIT"
 mockall = {version = ">= 0.11, <= 0.13", optional = true}
+rtactor-macros = { version="0.5.0", path = "rtactor-macros" }
+smol = { version = "2.0", optional = true }
+smol-macros = { version = "0.1", optional = true }
```
We don't need smol-macros for the lib code.

If we move it to the dev dep, I would avoid issue with smol >= 1.0 vs smol-macros < 1.0 dependency hell for the lib users.

stefanhp replied:
Fixed in 8fa6a45ca02480d476d758bb8afdf5455a8055b7

mrcmry commented on tests/async_actor.rs:
```diff
@@ -0,0 +1,94 @@
+}
+
+#[apply(test!)]
+/// Send a request from an active actor and use AsyncActor::responds() to respond.
```
```suggestion
/// Send a request from an async actor and use AsyncActor::responds() to respond.
```

mrcmry commented on src/async_actor.rs:
```diff
@@ -0,0 +1,365 @@
+use super::actor::{self, Message, RequestId};
```
I think we should rename `AsyncActor` -> `AsyncMailbox`. This is much more semantically correct. `async_actor.wait_message().await` is strange or plainly wrong semantically. `async_mailbox.wait_message().await`  make totally sense.

We can also optionally rename ActiveActor -> ActiveMailbox with a alias with deprecation for compatibility.

stefanhp replied:
Fixed in 8fa6a45ca02480d476d758bb8afdf5455a8055b7 and 4d6d8536b8f75439d2f1d716013751fe3d4cf8f3

mrcmry commented on src/async_actor.rs:
```diff
@@ -0,0 +1,365 @@
+
+////////////////////////////// public macros /////////////////////////////////////
+
+/// Create a struct that allows a blocking access to a address and implements one or more SyncRequester or SyncNotifier
```
```suggestion
/// Create a struct that allows a blocking access to a address and implements one or more AsyncRequester or AsyncNotifier
```

mrcmry commented on src/async_actor.rs:
````diff
@@ -0,0 +1,365 @@
+///
+/// Example:
+/// ```rs
+/// define_syn_accessor(MySyncAccessorStructName, FirstSyncRequesterName, SecondSyncRequesterName, ...);
````
```suggestion
/// define_async_accessor(MyAsyncAccessorStructName, FirstAsyncRequesterName, SecondAsyncRequesterName, ...);
```

mrcmry commented on src/async_actor.rs:
````diff
@@ -0,0 +1,365 @@
+/// ```rs
+/// define_syn_accessor(MySyncAccessorStructName, FirstSyncRequesterName, SecondSyncRequesterName, ...);
+/// ```
+/// The struct implements `SyncAccessData`. It will have a function new(&Addr) -> MySyncAccessorStructName
````
```suggestion
/// The struct implements `MyAsyncAccessorStructName`. It will have a function `new(&Addr) -> MyAsyncAccessorStructName`
```

mrcmry commented on src/async_actor.rs:
```diff
@@ -0,0 +1,365 @@
+            return Result::Ok(message);
+        }
+
+        // wait in mpsc queue
```
```suggestion
        // wait on the queue
```

mrcmry reviewed (changes requested):
I did an interim review after looking how async really works in rust.

You can integrate my remarks, discuss them in github or during the next daily.

This PR should stay in draft until we implement the TODO for delays.

stefanhp committed:
refactor: rename AsyncActor to AsyncMailbox

- fix typos
- remove smol-macros from dependencies (only for dev-dependencies)
- fix some spelling config

stefanhp committed:
refactor: rename ActiveActor to ActiveMailbox

- keep ActiveActor as a deprecated type alias for now

stefanhp committed:
fix typo and dead code

stefanhp committed:
build: remove unused dependency for lib

stefanhp committed:
feature: add smol and tokio features

- add smol and tokio features to support those async runtimes
- implement wait_for_rx_message with a timeout for each runtime

stefanhp committed:
Update Cargo.toml

Co-authored-by: mrcmry <marc.emery@amnis.ch>

stefanhp committed:
Merge pull request #12 from rtreactive/feature/tokio

Tokio and smol features

stefanhp committed:
doc: add async information in README

stefanhp committed:
build: update change log

mrcmry commented on src/async_actor.rs:
```diff
@@ -0,0 +1,412 @@
+    }
+
+    /// Send a request and wait for the response for a given duration.
+    // TODO implement timeout
```
```suggestion
```

stefanhp replied:
Removed in e7213bdda3e145db6b27710592de8bada10480b8

mrcmry commented on src/dispatcher.rs:
```diff
@@ -75,15 +75,15 @@ pub enum Request {
 /// An accessor for a dispatcher with specialized functions.
 pub struct SyncAccessor {
-    active_actor: ActiveActor,
+    active_actor: ActiveMailbox,
```
It would be nice to rename this active_maibox.

stefanhp replied:
Fixed in 1e7af0f45824f560cbc6d60f19b8eed7ee080b81

mrcmry commented on src/simulation.rs:
```diff
@@ -143,7 +143,7 @@ impl SimulationDispatcher {
     }
     /*
-        fn wait_active_message(&mut self, active_actor: &mut ActiveActor, timeout: Duration) -> Result<Message, actor::Error>
+        fn wait_active_message(&mut self, active_actor: &mut ActiveMailbox, timeout: Duration) -> Result<Message, actor::Error>
```
also nice to rename to active_mailbox or active

stefanhp replied:
Fixed in 1e7af0f45824f560cbc6d60f19b8eed7ee080b81

mrcmry commented on tests/active.rs:
```diff
@@ -9,16 +9,27 @@ fn test_addr() {
 
 #[test]
-/// Check if it's possible to send a buffer and extract it from a immutable Message reference.
+#[allow(deprecated)]
```
This test could be rewritten without `deprecated` like this:
```
fn test_deprecated_active_mailbox_addr() {
    let mut addr = Addr::new();
    assert!(!addr.is_valid());

    let mailbox = rtactor::ActiveMaibox::new(1);
    addr = mailbox.addr();
    assert!(addr.is_valid());
}

stefanhp replied:
No: the goal here is to test that the deprecated type alias is actually working...

mrcmry reviewed (approved):
Nice !

stefanhp committed:
refactor: remame attributes of type ActiveMailbox

- remame attributes of type ActiveMailbox to mailbox

stefanhp committed:
refactor: remove invalid comment

stefanhp committed:
test: fix sporadic test failure

When using timeouts in async functions, we can no longer use Duration::ZERO because in some cases a timeout of zero will actually timeout before the async call.

So changed to 500 millis.
